prov/gni/src/gnix_wait.c

```diff
@@ -30,11 +30,164 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  * SOFTWARE.
  */
-
+#include <stdlib.h>
+#include <signal.h>
 #include "gnix.h"
 #include "gnix_wait.h"
 
 /*******************************************************************************
+ * Lets create some threads for progress functions, if FI_PROGRESS AUTO is
+ * enabled we don't need to do this, in reality we are using the same functions
+ * we just don't want to have to pass weird amounts of information to the
+ * function, easier to spawn the thing and cancel it than do other things.
+ */
+
+static pthread_t	gnix_wait_thread;
+static pthread_mutex_t	gnix_wait_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t	gnix_wait_cond;
+static volatile int	gnix_wait_thread_ready;
+static volatile int	gnix_wait_thread_enabled;
+static volatile int	gnix_wait_thread_busy;
+static volatile int	gnix_wait_refcnt;
+
+/*
+ * It is necessary to have a separate thread making progress in order for the
+ * wait functions to succeed. Currently the AUTO_PROGRESS model used by gni
+ * does not check just make progress. This version of that thread is designed
+ * to always make progress so we don't hard stall while sitting on fi_wait.
+ * this function is intended to be invoked as an argument to pthread_create,
+ */
+static void *__gnix_wait_nic_prog_thread_fn(void *the_arg)
+{
+	int ret = FI_SUCCESS, prev_state;
+	struct gnix_nic *nic1, *nic2;
+	sigset_t  sigmask;
+	gni_return_t status;
+	gni_cq_entry_t cqe;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/*
+	 * temporarily disable cancelability while we set up
+	 * some stuff
+	 */
+
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &prev_state);
+
+	/*
+	 * help out Cray core-spec, say we're not an app thread
+	 * and can be run on core-spec cpus.
+	 */
+	gnix_wait_thread_ready = 1;
+	ret = _gnix_task_is_not_app();
+	if (ret)
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			"_gnix_task_is_not_app call returned %d\n",
+			ret);
+
+	/*
+	 * block all signals, don't want this thread to catch
+	 * signals that may be for app threads
+	 */
+
+	memset(&sigmask, 0, sizeof(sigset_t));
+	ret = sigfillset(&sigmask);
+	if (ret) {
+		GNIX_WARN(FI_LOG_EP_CTRL,
+		"sigfillset call returned %d\n", ret);
+	} else {
+
+		ret = pthread_sigmask(SIG_SETMASK,
+					&sigmask, NULL);
+		if (ret)
+			GNIX_WARN(FI_LOG_EP_CTRL,
+			"pthread_sigmask call returned %d\n", ret);
+	}
+
+	/*
+	 * okay now we're ready to be cancelable.
+	 */
+
+	pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, &prev_state);
+
+	pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
+
+try_again:
+	pthread_mutex_lock(&gnix_wait_mutex);
+	if (!gnix_wait_thread_enabled)
+		pthread_cond_wait(&gnix_wait_cond, &gnix_wait_mutex);
+	pthread_mutex_unlock(&gnix_wait_mutex);
+	status = GNI_RC_SUCCESS;
+	pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+	dlist_for_each_safe(&gnix_nic_list, nic1, nic2, gnix_nic_list) {
+		do {
+			status = GNI_CqGetEvent(nic1->rx_cq_blk,
```

Comment:
For 1: this clones the regular progress thread to a point, so the get_Cq_event comes from the other auto_progress_thread

For 2: I missed that, I didnt think a nic with references, but since this covers the entire nic list, some of the nics might disappear.  There is a gnix_nic_list_lock that covers this list I'll check into the feasibility of taking it while progressing the nic's, otherwise I might make an adjustment to the existing progress threads such that they switch to a usleep model for the duration of the fi_wait call. 

For 3: The other locations where _gnix_nic_progress is called under for the cntr_progress and for cq_progress utilized a dlist_for_each_safe just for calling _gnix_nic_progress. I based the need for the dlist_for_safe based on those usages.